Part 8 refactors.

HexDecimal opened the pull request:
My refactors of part 8 as I write them, so far I've merged the part 7 changes.

You could fork from this if you wanted to add some changes.

TStand90 commented:
I wrote the "part-8-experimenting" branch before merging any of your changes for parts 6 and 7, so a lot of the branch might be downright unusable. If you find that's the case, feel free to completely disregard it.

HexDecimal commented:
I've already merged the part-7 branch and with the exceptions of some turn handling bugs it works as intended.  I'll be replacing most of it with new code which will give the same results as the old code, but will be more abstract internally.

HexDecimal commented:
I made a new branch `part-8` copied from `part-7` and set that as the base.  The diffs and commits for this PR are now readable.

HexDecimal committed:
Remove duplicated action code.

HexDecimal committed:
Move event loop outside of the input handler class.

The event handler can change in response to an event. If the event loop
continues within a handler after it has switched to another then that
will cause bugs.  So the event loop MUST be outside of the handler.

HexDecimal committed:
Remove register function.

Not sure what this is for.  If this gets put back in it should be
implemented as a class decorator.

HexDecimal committed:
Merge upstream part-8 changes.

HexDecimal committed:
Extend hierarchical relations between GameMap's and component classes.

Now the gamemap of items are in sync if their container moves across
gamemaps.

BaseComponent removed from BaseAI.  This class had its own system setup
and doesn't need the features from BaseComponent.
(Also I'd have to rename a lot of attributes.)

HexDecimal committed:
Use a more restrictive Union.

HexDecimal committed:
Merge upstream changes.

HexDecimal committed:
Separate input handlers into subclasses.

HexDecimal committed:
Adds a look handler and refactors user input handling.

HexDecimal committed:
Refactor exiting out of user input handlers.
